hovercontrol/src/homie.rs

```diff
@@ -0,0 +1,209 @@
+use crate::config::{get_mqtt_options, MqttConfig};
+use crate::controller::{
+    DEFAULT_MAX_SPEED, DEFAULT_SCALE, DEFAULT_SPRING_CONSTANT, MAX_MAX_SPEED, MAX_SCALE,
+    MAX_SPRING_CONSTANT, SPRING_CONSTANT_STEP,
+};
+use eyre::Report;
+use homie_device::{HomieDevice, Node, Property};
+use log::{error, trace};
+use messages::{Side, SpeedLimits};
+use tokio::runtime::Handle;
+
+const HOMIE_PREFIX: &str = "homie";
+const HOMIE_DEVICE_ID: &str = "hoverkite";
+const HOMIE_DEVICE_NAME: &str = "Hoverkite";
+
+pub struct Homie<'a> {
+    homie: Option<HomieDevice>,
+    handle: &'a Handle,
+}
+
+impl<'a> Homie<'a> {
+    pub fn connect_and_start(
+        handle: &'a Handle,
+        config: Option<MqttConfig>,
+    ) -> Result<Homie<'a>, Report> {
+        let homie = if let Some(config) = config {
+            Some(handle.block_on(make_homie_device(config))?)
+        } else {
+            None
+        };
+        Ok(Self { homie, handle })
+    }
+
+    pub fn send_target(&self, side: Side, target: i64) {
+        self.send_property(node_id(side), "target", target);
+    }
+
+    pub fn send_centre(&self, side: Side, centre: i64) {
+        self.send_property(node_id(side), "centre", centre);
+    }
+
+    pub fn send_position(&self, side: Side, position: i64) {
+        self.send_property(node_id(side), "position", position);
+    }
+
+    pub fn send_max_speed(&self, max_speed: SpeedLimits) {
+        self.send_property("general", "max_speed", max_speed.positive);
+        self.send_property("general", "min_speed", max_speed.negative);
+    }
+
+    pub fn send_spring_constant(&self, spring_constant: u16) {
+        self.send_property("general", "spring_constant", spring_constant);
+    }
+
+    pub fn send_scale(&self, scale: f32) {
+        self.send_property("general", "scale", scale);
+    }
+
+    pub fn send_battery_readings(
+        &self,
+        side: Side,
+        battery_voltage: u16,
+        backup_battery_voltage: u16,
+        motor_current: u16,
+    ) {
+        let node_id = node_id(side);
+        self.send_property(node_id, "battery_voltage", battery_voltage);
+        self.send_property(node_id, "backup_battery_voltage", backup_battery_voltage);
+        self.send_property(node_id, "motor_current", motor_current);
```

Comment:
slightly unfortunate that there's no way to send these as an atomic update or give them a timestamp (if we were reporting telemetry at uart-line-speed, and using it to do post-mortems, this might become significant. Probably fine at our current reporting interval)